torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,212 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+#ifdef ncclGroupStart
+#define NCCL_VERSION 2
+#endif
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, std::unordered_map<int, ncclComm_t> > _communicators;
+std::vector<ncclComm_t *> _all_communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto comm: _all_communicators) {
+    ncclCommDestroy(*comm);
+  }
+  Py_RETURN_NONE;
+}
+
+static std::unordered_map<int, ncclComm_t> _get_communicator(const int *devs, int ndevices) {
+  std::stringstream hash;
+  for (int i = 0; i < ndevices; i++) {
+    hash << devs[i] << ",";
+  }
+  if (_communicators.find(hash.str()) == _communicators.end()) {
+    ncclComm_t *comms = (ncclComm_t*) malloc(sizeof(ncclComm_t) * ndevices);
+    assert(comms);
```

Comment:
asserts are no-ops (setuptools adds `-DNDEBUG` unless you have a debug build of Python)